cpp/include/cugraph/graph_mask.hpp

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/handle.hpp>
+#include <raft/core/span.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cstdint>
+#include <limits>
+#include <optional>
+#include <vector>
+
+namespace cugraph {
+
+/**
+ * Compile-time fast lookup of log2(num_bits(mask_t)) to eliminate
+ * the log2 computation for powers of 2.
+ * @tparam mask_t
+ */
+template <typename mask_t>
+__host__ __device__ constexpr int log_bits()
+{
+  switch (std::numeric_limits<mask_t>::digits) {
+    case 8: return 2;
+    case 16: return 4;
+    case 32: return 5;
+    case 64: return 6;
+    default: return log2(std::numeric_limits<mask_t>::digits);
```

Comment:
If I am not mistaken, this will use https://en.cppreference.com/w/cpp/numeric/math/log2.

Will this work when std::numeric_limits<mask_t>::digits == 9?

log2(9) will be 3.xxx and if this gets implicitly converted to int, it will return 3. If this is the intention,

I think

`return static_cast<int>(floor(log2(std::numeric_limits<mask_t>::digits))); ` will be clearer.

Or if you're assuming that the size of mask_t is power of two, should we better add an assert statement to make the intention clear? (e.g. https://stackoverflow.com/questions/108318/how-can-i-test-whether-a-number-is-a-power-of-2).

Reply:
Really, I think we could drop the explicit log2 from this function altogether. It's just a little utility meant to provide a fast lookup for log2 of (a very limited set of) powers of two. It's also used almost exclusively on device and so I didn't want to perform strict input checking. I really think the function belongs in the detail namespace but I use it in the sampling utils so I exposed it as part of the masking public API. I'll think about this one a bit more.